src/ckmeans.js

```diff
@@ -80,77 +226,16 @@ function ckmeans(data/*: Array<number> */, nClusters/*: number */)/*: Array<Arra
     // within-cluster sum of squares. It's similar to linear regression
     // in this way, and this calculation incrementally computes the
     // sum of squares that are later read.
-
-    // The outer loop iterates through clusters, from 0 to nClusters.
-    for (var cluster = 0; cluster < nClusters; cluster++) {
-
-        // At the start of each loop, the mean starts as the first element
-        var firstClusterMean = sorted[0];
-
-        for (var sortedIdx = Math.max(cluster, 1);
-             sortedIdx < sorted.length;
-             sortedIdx++) {
-
-            if (cluster === 0) {
-
-                // Increase the running sum of squares calculation by this
-                // new value
-                var squaredDifference = Math.pow(
-                    sorted[sortedIdx] - firstClusterMean, 2);
-                matrix[cluster][sortedIdx] = matrix[cluster][sortedIdx - 1] +
-                    (sortedIdx / (sortedIdx + 1)) * squaredDifference;
-
-                // We're computing a running mean by taking the previous
-                // mean value, multiplying it by the number of elements
-                // seen so far, and then dividing it by the number of
-                // elements total.
-                var newSum = sortedIdx * firstClusterMean + sorted[sortedIdx];
-                firstClusterMean = newSum / (sortedIdx + 1);
-
-            } else {
-
-                var sumSquaredDistances = 0,
-                    meanXJ = 0;
-
-                for (var j = sortedIdx; j >= cluster; j--) {
-
-                    sumSquaredDistances += (sortedIdx - j) /
-                        (sortedIdx - j + 1) *
-                        Math.pow(sorted[j] - meanXJ, 2);
-
-                    meanXJ = (sorted[j] + (sortedIdx - j) * meanXJ) /
-                        (sortedIdx - j + 1);
-
-                    if (j === sortedIdx) {
-                        matrix[cluster][sortedIdx] = sumSquaredDistances;
-                        backtrackMatrix[cluster][sortedIdx] = j;
-                        if (j > 0) {
-                            matrix[cluster][sortedIdx] += matrix[cluster - 1][j - 1];
-                        }
-                    } else {
-                        if (j === 0) {
-                            if (sumSquaredDistances <= matrix[cluster][sortedIdx]) {
-                                matrix[cluster][sortedIdx] = sumSquaredDistances;
-                                backtrackMatrix[cluster][sortedIdx] = j;
-                            }
-                        } else if (sumSquaredDistances + matrix[cluster - 1][j - 1] < matrix[cluster][sortedIdx]) {
-                            matrix[cluster][sortedIdx] = sumSquaredDistances + matrix[cluster - 1][j - 1];
-                            backtrackMatrix[cluster][sortedIdx] = j;
-                        }
-                    }
-                }
-            }
-        }
-    }
+    fillMatrices(sorted, matrix, backtrackMatrix);
```

Comment:
The body of `fillMatrices` could technically be inlined here. I kept it in a separate function since it's more true to the original implementation, but this can be easily changed if preferred.